### ISSUE TITLE
Proposal for designing (or adopting) an ETL language and a parser as a DataVec facade

### ISSUE THREAD

**Issue by vlovsky** [deeplearning4j/DataVec#224]
!!!This proposal is in research phase!!! 

I would like to have this issue as a request for comments. Also, I offer my sincere apologies in advance , just in case I accidentally stepped on someone's toes. 

This proposal is written with the following goals in mind.

1. Increase user satisfaction in multiple areas. Few primary ones are lowering the learning curve, lowering cumulative cost of coding and maintenance of the derivative artifacts, increasing portability of the derivative work.
2. Do not intervene with the current DataVec design. So, that it still functions and can be used as Java API without modifications.
3. Learn on mistakes of existing frameworks and languages. Reuse or adopt a worthy language if it exists. This obviously increases chances of success in achieving goal no.1

TL;DR of the advantages and disadvantages of the proposed language in comparison with not having it.

Pros
1. Easy to relate and understand because it an intrinsic property of a DSL. It has much more affordable learning curve if any at all.
2. Simplified transfer/sharing of knowledge and derivative work. Using the ETL scripts vs Java code
3. Increased portability to other platforms - it is relatively easy to write a parser to insert it as a facade to another functionally compatible engine
4. Faster prototyping
???

Cons
1. Additional (optional, you don't use it - you don't need it) layer with minor overhead. Unless the framework is intended to _parse_ ETL instructions at >10K IPS this can be mitigated with trivial tools.
2.  Designing a good language is not easy. We've seen languages come and go. Existing DSLs need to be leveraged.
???

Three major functions of any ETL system are: extract, transform, load. I will use term ETL going forward as a reasonable approximation to refer to the proposed language. These function in an ETL system correspond to the following activities: read data from an IO device abstraction (whatever that might be, local file, web data stream, etc) according to a specified source format; transform this data into a different data structure according to another specified target format; write data out to another IO device (this is just an abstraction and can be anything again, from files, output streams, database tables, to INDArray and RDD). An ETL language in its turn must provide necessary instructions to guide this process, such as configuration of the IO devices or the algorithms of the data transformations. According to the [DL4J: ETL User Guide](https://deeplearning4j.org/etl-userguide#diagram-of-available-etl-paths), this all is principally available via functionality of the DataVec, ND4J.

At this juncture, we must remember who are the end-users of this functionality - data scientists, scientists from academia with degree in mathematics, physics, biology, etc. Most of them don't even care about Java. Learning Java is not their primary objective. This knowledge for them is merely a nuisance that can't be (can it?) avoided. They want to manipulate their data. Proposed ETL language is a higher level abstraction (DSL, or domain specific language if you will), the only purpose of which is to hide away Java instructions that are not essential for data processing and leave only essential elements.

What comprises a good DSL language? Arguably, most successful DSL languages were designed by the people on the floor. Those that know and work with the domain day in and day out. So, presumably a good DSL to work with data would come from fields where data is in abundance. Thankfully, industry has already accumulated a lot of those tools and the end-users have accumulated a lot of the necessary knowledge to work with them. Not only it would be a shame not to reuse it, the end-users are deterred by the proprietary languages since the knowledge and/or derivative work cannot be easily transferred. Proposal is to tap into the existing tools for data manipulation and data loading, and "appropriate" many necessary elements for the proposed language. There a few noteworthy tools such as SQL, Hive, Oracle SQL loader, DB2 Load, MATLAB export/import (and data manipulation in general)

Most complicated stuff is over. Lastly, there should be a language parser that will translate ETL language instructions into DataVec instructions, do appropriate variable bindings, etc. This has been done before and it isn't overly complicated. Optionally, depending if the language gets any traction or will be invoked directly by end-users, the framework could be extended with a language interpreter, so that the end-user don't even need to fire Eclipse or IDEA to change something in their ETL scripts. But most likely usage scenarios will require REST or other type of remote APIs.

In conclusion. Considering that the execution components are already functional and from my experience, I believe that adding a DSL for manipulating data to the DL4J stack is beneficial and is worth exploring. It is also worth exploring if the same level abstraction is beneficial with respect to the NN configuration API ("language") in DL4J. Since all discussed benefits are associated with the improved end-user interaction and user experience, it's obvious how they are related. A potential candidate for research in this respect could be [Net#](https://docs.microsoft.com/en-us/azure/machine-learning/machine-learning-azure-ml-netsharp-reference-guide). I'm aware of the proposal to use the DL4J as a backend engine for Keras.  This doesn't negate direct end-user interaction with the DL4J stack.

**Comment by tomthetrainer** [comment]
My thoughts..

## **SQL**
SQL defines multiple operations in a single statement, and SQL therefore gets dodgy when you try to chain things. Select * from (select * from(select * from))))

The SQL interpreter orders those as the optimizer decides. 

## **PIG**
Apache pig it comes closer to what you describe, a series of operations described as a series of operations. 

Something like apache pig? Where the flow of data is something like this. 

**One or more Load Statements.** 
A = LOAD 'data' AS (f1:int, f2:int, B:bag{T:tuple(t1:int,t2:int)});

Where A is a "bag" or a "MultiSet" a collection of Records. More accurately at this point it would be a description of that bag. 

**A series of creations of bags/collections/multiset/datasets from operations on previous bags.** 
X = FOREACH A GENERATE f1, f2, f1%f2;

**Then a final store or dump statement that triggers the execution?**
Store x into path. 

**In our case the final endpoint would be shipping to a NN**

 ## What is the endpoint? 

Is this clearly defined in DL4J? Meaning when input hits the NN for inference or training is it wrapped up in a self describing format? When dl4J goes from RecordReader to RecordReaderDataSetIterator the endpoint is DataSet, would that be the endpoint for the DSL you propose? Or does the output need to be wrapped in some Serialization protocol.

**Comment by vlovsky** [comment]
Thanks for your feedback. I started to work on a proof of concept using SQL*Loader syntax. It will be very simplistic, just enough to cover few DataVec basics.

I already went through a handful of languages and standards that are at least remotely represent loading and manipulating data. One standard ISO/IEC 9075-9:2008 (SQL/MED, and SQL extension) worth mentioning that is supported by PostgreSQL. Currently I have an affinity for two candidates ORACLE SQL*Loader and PIG Latin (funny you mentioned that).

Regarding **SQL**. Even though you can write embedded queries like that, no one practically does it, because there is a more potent way. What you have presented, easily translates into a single select statement. What you might have meant is what does one do, when data needs to come from multiple sources (tables in SQL world). Well, then you use JOIN. However, SQL*Loader syntax isn't like an SQL (although it contains similar features)

On the other hand, **PIG Latin** is imperative for the most part and offers substantial advantages over SQL when it comes to unstructured hierarchical data.

Here is an approximate illustration of how very simplistic data load is programmed in SQL*Loader and PIG:

_SQL*Loader:_
```
LOAD DATA INFILE 'sample.dat' INTO TABLE emp FIELDS TERMINATED BY ","
(hiredate INTEGER, deptno INTEGER, job CHAR, mgr CHAR, ename CHAR, empno INTEGER, sal INTEGER)
```

_PIG Latin:_
```
emp = LOAD 'sample.dat' USING PigStorage(',') AS 
(hiredate:int, deptno:int, job:chararray, mgr:chararray, ename:chararray, empno:int, sal:int);

```
For more advanced cases see [SQL*Loader](http://docs.oracle.com/database/121/SUTIL/GUID-8D037494-07FA-4226-B507-E1B2ED10C144.htm) and [PIG](https://pig.apache.org/docs/r0.7.0/piglatin_ref1.html) manuals.

Not too much different, aren't they? But the devil is in the advanced use cases. I don't know PIG enough to comment reliably, but from what I understand it doesn't allow to use "schema" with embedded transformations. One would handle this in the subsequent GENERATE statements. On the other hand SQL*Loader doesn't allow joins in schema such that columns a,b come from file A and c,d from file B. Premise is that this would happen after the fact via SQL. Before we go any further with language evaluation, I propose to concentrate on the principal questions: is it useful? would it benefit DL4J stack? is it in line with the strategical goals? and as you mentioned what is the end-goal?

**Comment by tomthetrainer** [comment]
My SQL example was meant to prove a point, treat it as psuedo code, yes the example can be written as a single join instead of subselects, but the subselects match the way the user thinks about the problem. My point is a single statement vs a definition of a multistep process. SQL and SQL loader are designed for the operation to be a single statement. 

We missed a few frameworks that are relevant. 

Scoobi, Scalding, Scrunch, Cascading, and the family of tools designed to treat a multi-step hadoop job as a data flow. 

Some Scoobi code. 
val users = TextInput.fromTextFile(inputDirectory)
val transformed = users.map{user =>
	user.split(",")(2).toInt * 100


persist(toTextFile(transformed, outputDirectory))

As for wether or not it is useful, we should have some engineers provide an opinion on that. 

DataVec is getting some work done that may go a few steps in this direction.

**Comment by tomthetrainer** [comment]
Some datavec documentation

https://deeplearning4j.org/etl-userguide

**Comment by agibsonccc** [comment]
@tomthetrainer this being useful isn't the question. I told him to submit the proposal.We need an intermediary format for datavec so we can allow other language support.

**Comment by vlovsky** [comment]
Again thanks for the feedback. I'm literally few hours from completing the proof of technical concept. I wish I had one more Saturday. @tomthetrainer No one knows how people think. I worked with SQL for more than two decades, therefore my opinion is biased. But I know many who don't feel awkward to actually join two tables. Especially if the queries are correlated. You can't do it with a simple "intermediate" variable, since the reference goes in both directions.

Regardless, I'm stressing second time that SQL*Loader is absolutely not like SQL. The features that are reused from SQL are very simple. Plus, I'm not insisting on any given language or even the level of adoption. You'll see it from the example in the POC. My hopes are only that we find something that is concise, easy to pick up and at the same time widely known.

Also, thanks for the link. Had it in the proposal :)

**Comment by agibsonccc** [comment]
Of note here: Most folks understand data frames. The problem I have with this abstraction is binary data. I'm wondering if we can create something a bit more general than that?

The TransformProcess allows us to take inputs for example and just specify a transform "step" which could easily work for images as well as audio.

**Comment by tomthetrainer** [comment]
Sounds good thanks for working on this.

**Comment by tomthetrainer** [comment]
Patched the code with new example

**Comment by vlovsky** [comment]
Pushed first cut to my new ["pet project"](https://github.com/vlovsky/DataVecSL). Unfortunately, I did't have much time to work on it. One thing from my plan that I didn't mange to cover is the date transformation from the basic DataVec example. Parser is fine, but the Conductor will choke on any inline transformation functions. This will have to wait for now. Hopefully I will have some time on Saturday. Yet, I think it provides some basis for thoughts even as is.

@agibsonccc still not quite sure I understand the concern with the images. Perhaps, If you could elaborate with one or two concrete use cases, it will be more clear. One thing that I was thinking was scenario like this. Data consists of two parts - "string" record data and for each record there is a source of binary data. String data file may contain something like this:
```
2016-01-01 17:00:00.000,USA,100.00,file:///images/image1.png
2016-01-01 18:03:01.256,FR,73.20,file:///images/image2.png
2016-01-03 02:53:32.231,USA,1621.00,file:///images/image3.png
2016-01-03 09:30:16.832,USA,43.19,file:///images/image4.png
2016-01-04 23:01:52.920,MX,159.65,file:///images/image5.png
2016-01-05 02:28:10.648,CAN,26.33,file:///images/image6.png
2016-01-05 10:15:36.483,USA,-0.90,file:///images/image7.png
```
In this case the script might look like the following:
```
load data
from 'string_data.file'
delimited by ','
as (
    DateTimeString string,
    MerchantCountryCode enum("USA","CAN"),
    TransactionAmountUSD double,
    imageSource string filler, rem this will be removed from the final result
    image blob LoadImage(imageSource) terminated by EOF
);
```

**Comment by vlovsky** [comment]
Why this was closed?

**Comment by agibsonccc** [comment]
I think tom mis clicked?

**Comment by agibsonccc** [comment]
@vlovsky so this looks great - my main concern with images and dataframes in general is it doesn't understand what a "blob" is. You're hitting this concern. Spark folks typically try to say: "Select * from pixel" which doesn't work. I was just broadly complaining about the mistakes folks with dataframes make.

**Comment by vlovsky** [comment]
I can make any data type, as long as DataVec supports it. As far as
official syntax is concerned, it might support custom data types... not
sure. For simplicity and clarity sake I made a few deviations from the
official syntax anyways. For unknown to me reason it doesn't support if
expressions. It would instead require a custom function like:

f1 integer expression "custom_function(:f1)"

instead, I allowed something like

f1 integer (if f1 < 0 then 0 else f1)

**Comment by tomthetrainer** [comment]
Hah, sorry for closing this

**Comment by vlovsky** [comment]
After a few days of scavenging, stumbled upon these [two standards - PMML/PFA](http://dmg.org/) Seems promising on the outside, but haven't seen many implementations. Have you heard about them? What is your opinion? For reference, here are [few tutorials](http://dmg.org/pfa/docs/tutorial1/) and what seems like a [practical guide](http://www.kdd.org/kdd2016/papers/files/adp0884-pivarskiAcb.pdf) to deployment.

The precursor to this finding was [Cascading](http://www.cascading.org/projects/pattern/) extension Scalding.

**Comment by agibsonccc** [comment]
None of these are going to be relevant. These formats are known not to represent modern pipelines at all. Cascading and scalding really only deal with sql type data (this is the problem with all current ETL in general)

**Comment by vlovsky** [comment]
What a shame

**Comment by agibsonccc** [comment]
@vlovsky I wouldn't worry too much about it. Directly adapting the transform process as a simple pipeline makes a ton of sense to me. What *has* been robust is the EIP abstraction:
http://camel.apache.org/

Why not look closer at this?

**Comment by vlovsky** [comment]
@agibsonccc I had been involved with EIP for the last 6 - 7 years or so. Not Camel, but a proprietary software. Along with the BPMN and UML, EIP served us as a way to outline the high to mid level architecture for communicating between multiple modern and legacy components. Both are meant to be used as tools for visualization, i.e. fast prototyping, sketching of the integration and orchestration designs.

So, while I'm typing this I have a mixed understanding of what your affinity to the EIP is with respect to a language for describing data flows and transformations. Are you thinking of the composeability of the EIP patterns? Visual modeling? Or something else?

If you haven't already, I recommend to listen to [this summary](https://www.youtube.com/watch?v=Xwi1DU6KoQ4) by Gregor Hohpe of the EIP. Perhaps, one of us is taking EIP for something else. Please, share your opinion.

**Comment by agibsonccc** [comment]
My main affinity for it is simple: Data pipelines inherently have input uris and output uris. You then do a series of transformations from those uris which are your sources you then write to.

Datavec works as follows:
data type -> input format output format
a series of transforms over a given raw data source of a certain kind of input format

I've actually been a camel user for years, I'm more than familiar with the pattern which is why I bring it up.
The component I'm thinking of is connecting data sources together converging on a series of transformations which is essentially what transformprocess is.

What I'd like to see is a sort of dataflow spec language that can handle binary data *and* columnar which I think EIP gets close to given that it understands InputStreams as well as things like transformations over CSVs and text data.

**Comment by vlovsky** [comment]
Interesting. So, you think about it as some sort of orchestration (from SOA
vocabulary) if you will. Right? And you most likely want it to allow
re-usability and composeability of the atomic units of the configuration.
Let me digest this.

**Comment by vlovsky** [comment]
After additional consideration and revisiting the EIP. I composed my [idea here](https://docs.google.com/presentation/d/16s6Iq3JAMVDB7UY8CMJ9eF2DzUqJZN_P91CNRwkvkeQ/edit?usp=sharing)

Thoughts?

**Comment by huitseeker** [comment]
I'm ok with this in the principle — an API /DSL is sorely needed. But experience proves SQL is a bit too limited, sp. w.r.t aggregative operations (folds / catamorphisms, what have you). I think a superset of the Spark Dataset API (where the superset part is precisely the graceful degradation in cases with uneven # of features / sequence columns) would work well.

**Comment by vlovsky** [comment]
I was evaluating SQL as one of the base lexical building blocks. It translates quite well into the pipeline pattern which is also aligned with EIP. Concrete node functionality may be adopted from Spark API.
